### PR TITLE
Implement redis to the project

### PR DESCRIPTION
**Pull request by OleksandRomaniuk**

### CODE REVIEW

File: web/src/main/kotlin/com/example/ajaxproject/service/PrivateChatServiceImpl.kt [hunk 1]
```diff
@@ -43,13 +45,16 @@ class PrivateChatServiceImpl(
...
     }
 
-
     override fun getPrivateRoom(roomId: String): Mono<PrivateChatRoom> {
-        return privateChatRoomRepository.findChatRoomById(roomId)
-            .switchIfEmpty(Mono.error(NotFoundException("Room with $roomId not found")))
+        return redisPrivateChatRoomRepository.findById(roomId)
+            .switchIfEmpty(
+                privateChatRoomRepository.findChatRoomById(roomId)
+                    .flatMap { redisPrivateChatRoomRepository.save(it) }
```
Comment (ajax-gnatiuk-s) [inline web/src/main/kotlin/com/example/ajaxproject/service/PrivateChatServiceImpl.kt:55]:
```suggestion
                privateChatRoomRepository.findChatRoomById(roomId)
                    .flatMap { redisPrivateChatRoomRepository.save(it) }
```

File: web/src/main/kotlin/com/example/ajaxproject/repository/RedisPrivateChatRoomRepository.kt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.example.ajaxproject.repository
+
+import com.example.ajaxproject.model.PrivateChatRoom
+import reactor.core.publisher.Mono
+
+interface RedisPrivateChatRoomRepository {
```
Comment (ajax-gnatiuk-s) [inline web/src/main/kotlin/com/example/ajaxproject/repository/RedisPrivateChatRoomRepository.kt:6]: Don't specify infrastructure details inside an interface name. 

`Redis` is infrastructure detail.

`RedisPrivateChatRoomRepository` -> `CacheableRoomRepository` as example.

File: web/src/main/kotlin/com/example/ajaxproject/model/PrivateChatRoom.kt [hunk 3]
```diff
@@ -1,12 +1,15 @@
 package com.example.ajaxproject.model
 
-import org.springframework.data.annotation.Id
+import com.fasterxml.jackson.annotation.JsonProperty
 import org.springframework.data.mongodb.core.mapping.Document
 
 @Document("private-chat-room")
 data class PrivateChatRoom(
-    @Id
+    @JsonProperty("id")
```
Comment (ajax-gnatiuk-s) [inline web/src/main/kotlin/com/example/ajaxproject/model/PrivateChatRoom.kt:8]: Mongo model is about working with MONGO. Why do you need this annotation?

**Review by ajax-gnatiuk-s — CHANGES_REQUESTED** [review]

File: web/src/main/kotlin/com/example/ajaxproject/model/PrivateChatRoom.kt (hunk 3 above)
Comment (OleksandRomaniuk) [reply web/src/main/kotlin/com/example/ajaxproject/model/PrivateChatRoom.kt]: Fixed

**Comment by OleksandRomaniuk** [comment]
Rewrite structure and add test for it

File: web/src/test/kotlin/com/example/ajaxproject/PrivateRoomRepositoryIT.kt [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        // WHEN
+        cacheableRepository.save(newRoom).block()
+
+        // THEN
+        val cachedRoom = redisTemplate.opsForValue().get(roomId).block()
+
+        Assertions.assertNotNull(cachedRoom)
+        Assertions.assertEquals(newRoom, cachedRoom)
+    }
+}
```
Comment (ajax-gnatiuk-s) [inline web/src/test/kotlin/com/example/ajaxproject/PrivateRoomRepositoryIT.kt:67]: As for me it's better to join these two test functions.

The only difference between them is assertion block

File: web/src/main/kotlin/com/example/ajaxproject/repository/cacheable/CacheablePrivateRoomRepository.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import com.example.ajaxproject.repository.redis.PrivateChatRoomRedisRepository
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.stereotype.Repository
+import reactor.core.publisher.Mono
+
+@Repository
+class CacheablePrivateRoomRepository<T>(
+    @Qualifier("mongoPrivateRoomRepository") private val privateChatRoomRepository: PrivateChatRoomRepository,
+    private val privateChatRoomRedisRepository: PrivateChatRoomRedisRepository
+) : CacheableRepository<T> {
```
Comment (ajax-gnatiuk-s) [inline web/src/main/kotlin/com/example/ajaxproject/repository/cacheable/CacheablePrivateRoomRepository.kt:14]:
```suggestion
) : CacheableRepository<PrivateChatRoom> {
```

File: web/src/main/kotlin/com/example/ajaxproject/repository/cacheable/CacheablePrivateRoomRepository.kt [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    override fun findById(id: String): Mono<T> {
+        return privateChatRoomRedisRepository.findById(id)
+            .switchIfEmpty(
+                privateChatRoomRepository.findPrivateChatRoomById(id)
+                    .flatMap { privateChatRoomRedisRepository.save(it) }
+                    .switchIfEmpty(
+                        Mono.error(NoSuchElementException("Can't get entity by id $id"))
+                    ))
```
Comment (ajax-gnatiuk-s) [inline web/src/main/kotlin/com/example/ajaxproject/repository/cacheable/CacheablePrivateRoomRepository.kt:33]:
```suggestion
                    )
         )
```

File: web/src/main/kotlin/com/example/ajaxproject/repository/cacheable/CacheablePrivateRoomRepository.kt [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+import com.example.ajaxproject.model.PrivateChatRoom
+import com.example.ajaxproject.repository.PrivateChatRoomRepository
+import com.example.ajaxproject.repository.redis.PrivateChatRoomRedisRepository
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.stereotype.Repository
+import reactor.core.publisher.Mono
+
+@Repository
+class CacheablePrivateRoomRepository<T>(
```
Comment (ajax-gnatiuk-s) [inline web/src/main/kotlin/com/example/ajaxproject/repository/cacheable/CacheablePrivateRoomRepository.kt:11]: Why you have parameterized the repository to `<T>` ?

File: web/src/test/kotlin/com/example/ajaxproject/PrivateRoomRepositoryIT.kt (hunk 4 above)
Comment (OleksandRomaniuk) [reply web/src/test/kotlin/com/example/ajaxproject/PrivateRoomRepositoryIT.kt:67]: Discused it

File: web/src/main/kotlin/com/example/ajaxproject/repository/cacheable/CacheablePrivateRoomRepository.kt (hunk 7 above)
Comment (OleksandRomaniuk) [reply web/src/main/kotlin/com/example/ajaxproject/repository/cacheable/CacheablePrivateRoomRepository.kt:11]: discussed it

**Review by ajax-gnatiuk-s — APPROVED** [review]

**Pull request merged**